mesonbuild/compilers/compilers.py

```diff
@@ -689,6 +691,29 @@ def run(self, code: 'mesonlib.FileOrString', env: 'Environment', *,
             dependencies: T.Optional[T.List['Dependency']] = None) -> RunResult:
         raise EnvironmentException('Language %s does not support run checks.' % self.get_display_language())
 
+    # Caching run() in general seems too risky (no way to know what the program
+    # depends on), but some callers know more about the programs they intend to
+    # run.
+    def cached_run(self, code: str, env: 'Environment', *,
+                   extra_args: T.Union[T.List[str], T.Callable[[CompileCheckMode], T.List[str]], None] = None,
+                   dependencies: T.Optional[T.List['Dependency']] = None) -> RunResult:
+        run_check_cache = env.coredata.run_check_cache
+        args = self.build_wrapper_args(env, extra_args, dependencies, CompileCheckMode('link'))
+        key = (code, tuple(args))
+        if key in run_check_cache:
+            p = env.coredata.run_check_cache[key]
```

Comment:
We saved a reference to this without `env.coredata`, why are we now using both versions?

Reply:
I had a fight with mypy earlier, and while trying to find the issue, I ended up trying to use the "full path". But it obviously was something else. I'll fix it in the next version.

Reply:
I think you need to define something similar to CompilerCheckCacheKey, and do `key: coredata.CompilerRunCacheKey = (code, tuple(args))`

Reply:
The patch already has that:
```
    # code, args
    RunCheckCacheKey = T.Tuple[str, T.Tuple[str, ...]]

```

Reply:
Now removed the pointless use of `env.coredata.`.